### PR TITLE
ci: Enable KVM group perms

### PR DESCRIPTION
**Pull request by KazuCocoa**
Just curious about Ubuntu and Enable KVM group perms

https://github.blog/changelog/2024-04-02-github-actions-hardware-accelerated-android-virtualization-now-available/
https://github.com/reactivecircus/android-emulator-runner?tab=readme-ov-file#running-hardware-accelerated-emulators-on-linux-runners

https://github.com/appium/appium-uiautomator2-driver/actions/runs/8012939666

Maybe this is the most recent succeeded cases:

Screenshot 2024-05-01 at 11 09 46 PM

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
We could probably use https://github.com/android-actions/setup-android, so doctor checks succeed

**Comment by KazuCocoa** [comment]
oh... they took 3h+. Hm, interesting

**Comment by KazuCocoa** [comment]
Worked but 3 hours right now.
https://github.com/appium/appium-uiautomator2-driver/actions/runs/8929436868/job/24527220932

Screenshot 2024-05-02 at 5 24 24 PM

It's better than nothing, but we could probably tune them.

**Comment by mykola-mokhnach** [comment]
We may also try to run API 30+ test on macOS as it is based on ARM now. I assume it might be faster in comparison intel/kvm

File: .github/workflows/functional-test.yml [hunk 1]
```diff
@@ -44,29 +44,29 @@ jobs:
...
-        brew install xq
         npm install mocha-multi-reporters --save-dev
         npm install -g appium
         npm install --no-save mjpeg-consumer
       name: Install dev dependencies
-    - uses: actions/setup-java@v3
-      with:
-        distribution: 'temurin'
-        java-version: '11'
+    # - uses: actions/setup-java@v3
```
Comment (mykola-mokhnach) [inline .github/workflows/functional-test.yml:60]: I assume we should install java 17

File: .github/workflows/functional-test.yml [hunk 2]
```diff
@@ -44,29 +44,29 @@ jobs:
...
+    #   with:
+    #     distribution: 'temurin'
+    #     java-version: '11'
     - run: |
         cwd=$(pwd)
         pushd "$cwd"
         cd ~
         CHROMEDRIVER_VERSION="${{ matrix.chromedriverVersion }}" appium driver install --source=local "$cwd"
-        appium driver doctor uiautomator2
+        # appium driver doctor uiautomator2
```
Comment (mykola-mokhnach) [inline .github/workflows/functional-test.yml:69]: we need to install android SDK before running doctor

**Comment by mykola-mokhnach** [comment]
I've tried to run emulator on macos14 and faced the same issue as in https://github.com/rust-minidump/minidump-writer/pull/116
I don't think we would be able to use emulator with newer mac version for now

**Comment by KazuCocoa** [comment]
Screenshot 2024-05-03 at 7 27 25 AM

File: .github/workflows/functional-test.yml [hunk 3]
```diff
@@ -44,23 +44,38 @@ jobs:
...
-        java-version: '11'
+        java-version: '17'
+    - name: Enable KVM group perms
+      run: |
+        echo 'KERNEL=="kvm", GROUP="kvm", MODE="0666", OPTIONS+="static_node=kvm"' | sudo tee /etc/udev/rules.d/99-kvm4all.rules
+        sudo udevadm control --reload-rules
+        sudo udevadm trigger --name-match=kvm
+    - name: Setup Android SDK
+      uses: android-actions/setup-android@v3
+    - name: AVD cache
```
Comment (mykola-mokhnach) [inline .github/workflows/functional-test.yml:71]: Caching actually needs two steps. Check the last example at https://github.com/ReactiveCircus/android-emulator-runner?tab=readme-ov-file#usage--examples

File: .github/workflows/functional-test.yml [hunk 4]
```diff
@@ -44,23 +44,38 @@ jobs:
...
       with:
         distribution: 'temurin'
-        java-version: '11'
+        java-version: '17'
+    - name: Enable KVM group perms
+      run: |
+        echo 'KERNEL=="kvm", GROUP="kvm", MODE="0666", OPTIONS+="static_node=kvm"' | sudo tee /etc/udev/rules.d/99-kvm4all.rules
+        sudo udevadm control --reload-rules
+        sudo udevadm trigger --name-match=kvm
+    - name: Setup Android SDK
```
Comment (mykola-mokhnach) [inline .github/workflows/functional-test.yml:69]: we also won't need this step if we add the emulator caching step before

Comment (KazuCocoa) [reply .github/workflows/functional-test.yml:69]: hm, when i removed this task, https://github.com/appium/appium-uiautomator2-driver/actions/runs/8945474040/job/24574511963 failed. Perhaps should we keep this...?

**Comment by KazuCocoa** [comment]
one interesting thing is the api level 32 one significantly ran fast in the last run:

Screenshot 2024-05-03 at 7 16 05 PM

File: .github/workflows/functional-test.yml (hunk 4 above)
Comment (mykola-mokhnach) [reply .github/workflows/functional-test.yml:69]: hm, yes, it makes sense if the avd preparation step is skipped

**Review by mykola-mokhnach — APPROVED** [review]

**Pull request merged**